src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java

```diff
@@ -297,15 +298,66 @@ public SegmentMetadata.ComponentMetadataMap writeData(IndexDescriptor indexDescr
                 if (vectorPostings.shouldAppendDeletedOrdinal())
                     deletedOrdinals.add(vectorPostings.getOrdinal());
             }
+
+            // map of existing ordinal to rowId (aka new ordinal if remapping is possible)
+            BiMap <Integer, Integer> ordinalMap = HashBiMap.create();
+            boolean canFastFindRows = false;
+            if (deletedOrdinals.size() == 0) {
+                canFastFindRows = true;
+                int minRow = Integer.MAX_VALUE;
+                int maxRow = Integer.MIN_VALUE;
+                for (VectorPostings<T> vectorPostings : postingsMap.values())
+                {
+                    if (vectorPostings.getRowIds().size() != 1)
+                    {
+                        canFastFindRows = false;
+                        ordinalMap.clear();
+                        break;
+                    }
+                    int rowId = vectorPostings.getRowIds().getInt(0);
+                    int ordinal = vectorPostings.getOrdinal();
+                    minRow = Math.min(minRow, rowId);
+                    maxRow = Math.max(maxRow, rowId);
+                    if(rowId != ordinal)
+                    {
+                        if (ordinalMap.containsKey(ordinal) && ordinalMap.get(ordinal) != rowId)
```

Comment:
this is more complex than it needs to be, since each ordinal should only be present once in postingsMap.  Just do `assert !ordinalMap.containsKey(ordinal)` and do the `put`

Reply:
asserts are not normally executed in production (requires "-ea" for jvm, and enabled for debug or IDE adds "-ea" for debug). asserts should not be used for arguments validation of production code.
I'd prefer to not write the marker that canFastFindRows works in this case.
We can do a preconditions check / throw exception (I don't know if it is a good idea to throw in this case, I think there is a potential data loss/failure to write).

Reply:
this isn't argument validation, this is "check to make sure the state matches what we think" which is exactly what assert is for